### ISSUE TITLE
Nodes Hight

### ISSUE THREAD

**Issue by Mayk-CLP** [KillBait/AutoDrive_Course_Editor#15]
Hi.

If I select "Fix Node Hight" or do a new connecting between to existing nodes they are sometimes too high.

**Comment by KillBait** [comment]
Can you elaborate a bit further on "too high"?, are they so high you cannot access them? or just higher than the other nodes?

The only heightmap data the editor has access to is not very accurate, so it can never place them 100% accurate (if it does, that is a happy coincidence).

**Comment by BattleHook** [comment]
I noticed this too, but the nodes still work and are accesable. If you ant to have them at exact ground level you have to LMB=select in the editor **in-game**.
Other issue though is nodes on viaduct and bridges. These nodes end up under the pavemant. A possible solution is maybe to be able to select a node and give it the same height as another selected node? like aligning horizontal/vertical, but then in the "_height-plane_" ?

**Comment by KillBait** [comment]
The node placement height isn't ideal, but it's the best i can do given the heightmap data, i am still looking for another way to
make the height more accurate, with some luck, that might improve.

A vertical alignment is doable, it shouldn't pose much of an issue.

**Comment by Mayk-CLP** [comment]
I have the same issue with nodes under Bridges. I think you are right that the Hight Map is not good enough to make a difference between some Grass and the Street level.

**Comment by Dzi4d3k** [comment]
Maybe  assumption that neighboring points height shouldn't differ more than let say 1 meter would be enough?

**Comment by BattleHook** [comment]
@KillBait  
"Added a menu item 'Fix Nodes Height', it will scan all nodes and fix the height to the correct values
(if a heightmap was loaded) "

Some suggestions to improve : 
Maybe only perform "Fix Nodes Height" on selected nodes instead of on all nodes ?
Maybe give nodes with exact integer value a different symbol (e.g. triangle instead of round) ?

**Comment by KillBait** [comment]
> Maybe assumption that neighbouring points height shouldn't differ more than let say 1 meter would be enough?

My current thought is exactly this, but 1 meter is quite a lot, maybe a half meter is a good compromise.

> Maybe only perform "Fix Nodes Height" on selected nodes instead of on all nodes ?

This has already changed and will be in the next version, it will only adjust a node if its height is hugely different from the heightmap location (something like 10 meters), running all nodes was just overkill in hindsight.
> Maybe give nodes with exact integer value a different symbol (e.g. triangle instead of round) ?

If no heightmap is loaded and a new node is placed, it will default the Y to -1, is easy to change the node look to
point those out, 

the current list i have is.

(1) Add a height alignment like @BattleHook suggested.
(2) Add a manual dialog box to edit the nodes location ( either menu or button )
(3) making any connection to a blank node inherit the starting nodes height (if it's valid).

Feel free to suggest anything else.

**Comment by BattleHook** [comment]
@KillBait
I think that if those 3 items in your list get incorporated it is a huge improvement and I would be glad to see these incorporated.

**Comment by Dzi4d3k** [comment]
> My current thought is exactly this, but 1 meter is quite a lot, maybe a half meter is a good compromise.

In first place I was going to suggest 0,5m, but I wonder what is longest distance between waypoints, and how it look on sloppy but straight road. Note that in FS it's common that slopes are than 10%, which IRL is already much, so if distance between waypoints will be more than 10m then with max neighboring height = 1m it will be not possible to do 10% slope. However it will be possible to create more waypoints and cope with higher slopes.

**Comment by KillBait** [comment]
It's been my observation that the max spacing AD puts between nodes is ~12 meters when your going in a straight line, a curved route obviously reduces the spacing.

One thought just came to me, as connection nodes can auto generate in-between nodes, how would averaging the difference
between the start and end nodes work. for example, start node A has a y level 50, end node B has a y level 70, if 3 nodes are
generated in-between,  should the y levels be A=50, auto1=55, auto2=60, auto3=65, B=70. This sounds like a not terrible
solution seeing as heightmap info might not always be available.

**Comment by Dzi4d3k** [comment]
> start node A has a y level 50, end node B has a y level 70, if 3 nodes are generated in-between, should the y levels be A=50, auto1=55, auto2=60, auto3=65, B=70.

I would agree, cause I highly doubt that there will, but any bumps in between like on motocross track ;D

**Comment by alex2ruote** [comment]
This solution was my first thought too, but  I wasn't sure about what's doable. This editor is truly a fantastic tool compared to the early versions and the only real bug left is this, but it's sadly a big one. If the height problem is fixed, one could just place enough key waypoints ingame and then finish the work with the editor much faster than driving around for hours.

**Comment by KillBait** [comment]
> (1) Add a height alignment like @BattleHook suggested.
(2) Add a manual dialog box to edit the nodes location ( either menu or button )
(3) making any connection to a blank node inherit the starting nodes height (if it's valid).

i just finished adding all 3 of these, initial testing seems ok,  just need to try them a bit longer to see if they break

**Comment by KillBait** [comment]
I have push the commit and a .jar file to the beta branch, feel free to test it, there shouldn't be anything config breaking in it, but i recommend you use backup configs just in case

**Comment by alex2ruote** [comment]
Nice work, KillBait.
The height alignment works pretty good. The only thing that would make it better would be the ability to select two points instead of one to close distant gaps where the two points differ in height. For short distances and flat terrain it's already a big help though.

![Height Alignment](https://user-images.githubusercontent.com/48148748/148053834-18f3df53-e0bb-4fd6-a1d4-1c76baf3f11a.jpg)
